lib/srv/discovery/common/renaming.go

```diff
@@ -0,0 +1,238 @@
+	return makeDiscoverySuffix(fn, name, matcherType, region, accountID)
+}
+
+// makeAzureDiscoverySuffix makes a discovery suffix for Azure resources, of the
+// form <matcher type>-<region>-<resource group>-<subscription ID>.
+func makeAzureDiscoverySuffix(fn suffixValidatorFn, name, matcherType, region, resourceGroup, subscriptionID string) string {
+	return makeDiscoverySuffix(fn, name, matcherType, region, resourceGroup, subscriptionID)
+}
+
+// makeGCPDiscoverySuffix makes a discovery suffix for GCP resources, of the
+// form <matcher type>-<location>-<project ID>.
+func makeGCPDiscoverySuffix(fn suffixValidatorFn, name, matcherType, location, projectID string) string {
+	return makeDiscoverySuffix(fn, name, matcherType, location, projectID)
+}
+
+// applyDiscoveryNameSuffix takes a resource with labels and a suffix to add
+// to the name, then modifies the resource to add a label containing the
+// original name and sets a new name with the suffix appended.
+// This function does nothing if the suffix is empty.
+func applyDiscoveryNameSuffix(resource types.ResourceWithLabels, suffix string) {
+	if suffix == "" {
+		// nop if suffix parts aren't given.
+		return
+	}
+	discoveredName := resource.GetName()
+	labels := resource.GetStaticLabels()
+	if labels == nil {
+		labels = make(map[string]string)
+	}
+	// set the originally discovered name as a label.
+	labels[types.DiscoveredNameLabel] = discoveredName
+	resource.SetStaticLabels(labels)
+	// update the resource name with a suffix.
+	resource.SetName(fmt.Sprintf("%s-%s", discoveredName, suffix))
+}
+
+// suffixValidatorFn is a func that validates a suffix.
+type suffixValidatorFn func(string) error
+
+// databaseNameValidator is a suffixValidatorFn for databases.
+func databaseNameValidator(part string) error {
+	// validate the suffix part adding a simple stub prefix "a" and
+	// validating it as a full database name.
+	return types.ValidateDatabaseName("a" + part)
+}
+
+// kubeClusterNameValidator is a suffixValidatorFn for kube clusters.
+func kubeClusterNameValidator(part string) error {
+	// validate the suffix part adding a simple stub prefix "a" and
+	// validating it as a full kube cluster name.
+	return types.ValidateKubeClusterName("a" + part)
+}
+
+// makeDiscoverySuffix takes a list of suffix parts and a suffix validator func,
+// sanitizes each part and checks it for validity, then joins the result with
+// hyphens "-".
+func makeDiscoverySuffix(validatorFn suffixValidatorFn, name string, parts ...string) string {
+	// convert name to lower case for substring checking.
+	name = strings.ToLower(name)
+	var out []string
+	for _, part := range parts {
+		part = sanitizeSuffixPart(part)
+		// skip blank parts.
+		if part == "" {
+			continue
+		}
+		// skip redundant parts.
+		if strings.Contains(name, strings.ToLower(part)) {
+			continue
+		}
+		// skip invalid parts.
+		if err := validatorFn(part); err != nil {
```

Comment:
Do we need to validate suffix parts  ? 
The behaviour where part was not applied because is invalid and the `makeDiscoverySuffix` produce different output than expected can by misleading. 
For instance AWS accountID contains only numbers  so the `validatorFn(` 
```validDatabaseNameRegexp = regexp.MustCompile(`^[a-zA-Z]([-a-zA-Z0-9]*[a-zA-Z0-9])?$`)``` 
will return an error so this part will be skipped 

Could you add UT coverage ?

Reply:
the idea is to build a suffix that incorporates additional info, but skip parts that would violate the naming validation.

So suffix building is just best effort. This is especially a problem for Azure which allows all kinds of characters in resource group name etc.

Without this checking, we may prevent a database from being discovered due to our own renaming schemes.

Reply:
I'll unit test for this func though

Reply:
Note that in the RFD, I mentioned this issue, and offered an alternative to dropping the info:

>If resources exist within the Azure resource group weird-)(-group-name, then we simply drop the resource group name from the resource name:
>
>foo-eastus-aks-11111111-1111-1111-1111-111111111111
>foo-eastus-aks-22222222-2222-2222-2222-222222222222
>foo-eastus-mysql-11111111-1111-1111-1111-111111111111
>...
>Unfortunately, this would allow name collisions across resource groups.

>Alternatively, we could apply a transformation to the resource group name to make it valid. For example, base64 encode it, make the string lowercase, and replace the [+/=] characters with valid characters, maybe even truncating the result: (another heuristic, although less likely to collide names):
>$ echo "weird-)(-group-name" | base64 | sed 's#[+/=]#x#g' | tr '[:upper:]' '[:lower:]' | cut -c1-8 
d2vpcmqt
>$ echo "other-weird-)(-group-name" | base64 | sed 's#[+/=]#x#g' | tr '[:upper:]' '[:lower:]' | cut -c1-8 
b3rozxit
> making names that look like:
> foo-eastus-aks-d2vpcmqt-11111111-1111-1111-1111-111111111111
> foo-eastus-aks-b3rozxit-11111111-1111-1111-1111-111111111111

I'm still open to doing something like this instead of skipping invalid parts if you prefer.

Reply:
added unit testing for `makeDiscoverySuffix`: [c3b3822](https://github.com/gravitational/teleport/pull/28845/commits/c3b3822f0ceab6cb0b5ad63ca86cde2ab46dc825)

Reply:
Thanks. I miss that from AWS AccountID the `a` prefix is added to make the name valid before calling `ValidateResourceName` 
Can we consider renaming [databaseNameValidator](https://github.com/gravitational/teleport/blob/88d0c5c457f07a8af23369d181d59ea4e41f826d/lib/srv/discovery/common/renaming.go#L177) to `databasePartNameValidation`   ?

Reply:
4d93110b711c33dac79c3ae99da946e6142ee81d